components/profiler/src/profiler_linux.rs

```diff
@@ -0,0 +1,68 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::sync::Mutex;
+
+use callgrind::CallgrindClientRequest;
+
+#[derive(Debug, PartialEq)]
+enum Profiler {
+    None,
+    GPerfTools,
+    CallGrind,
+}
+
+lazy_static::lazy_static! {
+    #[derive(Debug)]
+    static ref ACTIVE_PROFILER: Mutex<Profiler> = Mutex::new(Profiler::None);
+}
+
+/// Start profiling.
+///
+/// When `profiling` feature is not enabled, this function will do nothing and there is totally
+/// zero cost.
+///
+/// When running in Callgrind, Callgrind instrumentation will be started
+/// (`CALLGRIND_START_INSTRUMENTATION`). Otherwise, the CPU Profiler will be started and profile
+/// will be generated to the file specified by `name`.
+// TODO: Better multi-thread support.
+#[inline]
+pub fn start(name: impl AsRef<str>) {
+    let mut profiler = ACTIVE_PROFILER.lock().unwrap();
+
+    // Profiling in progress.
+    if *profiler != Profiler::None {
+        return;
```

Comment:
Should it return error instead? For example, what if two profiling requests are received at the same time?